Modules/_zstd/compressor.c

```diff
@@ -45,98 +45,101 @@ typedef struct {
+    /* Save for generating ZSTD_CDICT */
+    self->compression_level = (int)level;
 
-        /* Set compressionLevel to compression context */
-        zstd_ret = ZSTD_CCtx_setParameter(self->cctx,
-                                          ZSTD_c_compressionLevel,
-                                          level);
+    /* Set compressionLevel to compression context */
+    const size_t zstd_ret = ZSTD_CCtx_setParameter(
+        self->cctx, ZSTD_c_compressionLevel, (int)level);
 
-        /* Check error */
-        if (ZSTD_isError(zstd_ret)) {
-            set_zstd_error(mod_state, ERR_SET_C_LEVEL, zstd_ret);
+    /* Check error */
+    if (ZSTD_isError(zstd_ret)) {
+        const _zstd_state* const st = PyType_GetModuleState(Py_TYPE(self));
+        if (st == NULL) {
             return -1;
         }
-        return 0;
+        set_zstd_error(st, ERR_SET_C_LEVEL, zstd_ret);
+        return -1;
     }
+    return 0;
+}
 
-    /* Options dict */
-    if (PyDict_Check(level_or_options)) {
-        PyObject *key, *value;
-        Py_ssize_t pos = 0;
+static int
+_zstd_set_c_parameters(ZstdCompressor *self, PyObject *options)
+{
+    /* Set options dict */
+    _zstd_state* const mod_state = PyType_GetModuleState(Py_TYPE(self));
+    if (mod_state == NULL) {
+        return -1;
+    }
 
-        while (PyDict_Next(level_or_options, &pos, &key, &value)) {
-            /* Check key type */
-            if (Py_TYPE(key) == mod_state->DParameter_type) {
-                PyErr_SetString(PyExc_TypeError,
-                                "Key of compression option dict should "
-                                "NOT be DecompressionParameter.");
-                return -1;
-            }
+    if (!PyDict_Check(options)) {
+        PyErr_Format(PyExc_TypeError, "invalid type for options, expected dict");
+        return -1;
+    }
 
-            int key_v = PyLong_AsInt(key);
-            if (key_v == -1 && PyErr_Occurred()) {
-                PyErr_SetString(PyExc_ValueError,
-                                "Key of options dict should be a CompressionParameter attribute.");
-                return -1;
-            }
+    Py_ssize_t pos = 0;
+    PyObject *key, *value;
+    while (PyDict_Next(options, &pos, &key, &value)) {
+        /* Check key type */
+        if (Py_TYPE(key) == mod_state->DParameter_type) {
+            PyErr_SetString(PyExc_TypeError,
+                            "key should NOT be DecompressionParameter.");
+            return -1;
+        }
 
-            // TODO(emmatyping): check bounds when there is a value error here for better
-            // error message?
-            int value_v = PyLong_AsInt(value);
-            if (value_v == -1 && PyErr_Occurred()) {
-                PyErr_SetString(PyExc_ValueError,
-                                "Value of option dict should be an int.");
-                return -1;
-            }
+        const int key_v = PyLong_AsInt(key);
+        if (key_v == -1 && PyErr_Occurred()) {
```

Comment:
Do not override arbitrary exceptions. Check the type of exception before replacing it.